asyncio_mode=auto does not support hypothesis without marker

The new `asyncio_mode=auto` mode does not support hypothesis markers. I'm assuming this is an issue on the asyncio side, since `strict` mode continues to work, but please let me know if I should post this to `hypothesis` instead!

I think @asvetlov and @seifertm are the relevant folks to ping for this.

Reproduction case:

<details>
<summary>pytest.ini</summary>

```
[pytest]
asyncio_mode = auto
```
</details>

<details>
<summary>test.py</summary>

```
from hypothesis import given
from hypothesis import strategies as st

@given(x=st.booleans())
async def test_foo(x):
    assert x
```
</details>

```
» pip install pytest-asyncio hypothesis
» pytest test.py
== test session starts =
platform darwin -- Python 3.7.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /private/tmp/testpyasynchypo, configfile: pytest.ini
plugins: hypothesis-6.35.0, asyncio-0.17.0
collected 1 item                                                                                                                                                                                                            

test.py F                                                                       [100%]

== FAILURES ==
_ test_foo _

    @given(x=st.booleans())
>   async def test_foo(x):
E   hypothesis.errors.InvalidArgument: Hypothesis doesn't know how to run async test functions like test_foo.  You'll need to write a custom executor, or use a library like pytest-asyncio or pytest-trio which can handle the translation for you.
E       See https://hypothesis.readthedocs.io/en/latest/details.html#custom-function-execution

test.py:6: InvalidArgument
== short test summary info ==
FAILED test.py::test_foo - hypothesis.errors.InvalidArgument: Hypothesis doesn't know how to run async test functions like test_foo.  You'll need to write a custom executor, or use a library like pytest-asyncio or pyte...
== 1 failed in 0.22s ==
```

Comparison, using `strict` mode, the following works as expected:

<details>
<summary>pytest.ini</summary>

```
[pytest]
asyncio_mode = strict
```
</details>

<details>
<summary>test.py</summary>

```
import pytest
from hypothesis import given
from hypothesis import strategies as st

@pytest.mark.asyncio
@given(x=st.booleans())
async def test_foo(x):
    assert x
```
</details>

--- a/README.rst
+++ b/README.rst
@@ -256,6 +256,10 @@ or an async framework such as `asynctest <https://asynctest.readthedocs.io/en/la
 
 Changelog
 ---------
+0.17.1 (UNRELEASED)
+~~~~~~~~~~~~~~~~~~~
+- Fixes a bug that prevents async Hypothesis tests from working without explicit ``asyncio`` marker when ``--asyncio-mode=auto`` is set. `#258 <https://github.com/pytest-dev/pytest-asyncio/issues/258>`_
+
 0.17.0 (22-01-13)
 ~~~~~~~~~~~~~~~~~~~
 - `pytest-asyncio` no longer alters existing event loop policies. `#168 <https://github.com/pytest-dev/pytest-asyncio/issues/168>`_, `#188 <https://github.com/pytest-dev/pytest-asyncio/issues/168>`_

--- a/pytest_asyncio/plugin.py
+++ b/pytest_asyncio/plugin.py
@@ -115,7 +115,13 @@ def pytest_configure(config):
 @pytest.mark.tryfirst
 def pytest_pycollect_makeitem(collector, name, obj):
     """A pytest hook to collect asyncio coroutines."""
-    if collector.funcnamefilter(name) and _is_coroutine(obj):
+    if not collector.funcnamefilter(name):
+        return
+    if (
+        _is_coroutine(obj)
+        or _is_hypothesis_test(obj)
+        and _hypothesis_test_wraps_coroutine(obj)
+    ):
         item = pytest.Function.from_parent(collector, name=name)
         if "asyncio" in item.keywords:
             return list(collector._genfunctions(name, obj))
@@ -126,6 +132,10 @@ def pytest_pycollect_makeitem(collector, name, obj):
                 for elem in ret:
                     elem.add_marker("asyncio")
                 return ret
+
+
+def _hypothesis_test_wraps_coroutine(function):
+    return _is_coroutine(function.hypothesis.inner_test)
 
 
 class FixtureStripper:

--- a/pytest_asyncio/plugin.py
+++ b/pytest_asyncio/plugin.py
@@ -288,7 +288,7 @@ def pytest_pyfunc_call(pyfuncitem):
     where the wrapped test coroutine is executed in an event loop.
     """
     if "asyncio" in pyfuncitem.keywords:
-        if getattr(pyfuncitem.obj, "is_hypothesis_test", False):
+        if _is_hypothesis_test(pyfuncitem.obj):
             pyfuncitem.obj.hypothesis.inner_test = wrap_in_sync(
                 pyfuncitem.obj.hypothesis.inner_test,
                 _loop=pyfuncitem.funcargs["event_loop"],
@@ -298,6 +298,10 @@ def pytest_pyfunc_call(pyfuncitem):
                 pyfuncitem.obj, _loop=pyfuncitem.funcargs["event_loop"]
             )
     yield
+
+
+def _is_hypothesis_test(function) -> bool:
+    return getattr(function, "is_hypothesis_test", False)
 
 
 def wrap_in_sync(func, _loop):

--- a/tests/hypothesis/test_base.py
+++ b/tests/hypothesis/test_base.py
@@ -2,6 +2,7 @@
 sync shim for Hypothesis.
 """
 import asyncio
+from textwrap import dedent
 
 import pytest
 from hypothesis import given, strategies as st
@@ -40,3 +41,48 @@ async def test_can_use_fixture_provided_event_loop(event_loop, n):
     semaphore = asyncio.Semaphore(value=0)
     event_loop.call_soon(semaphore.release)
     await semaphore.acquire()
+
+
+def test_async_auto_marked(pytester):
+    pytester.makepyfile(
+        dedent(
+            """\
+        import asyncio
+        import pytest
+        from hypothesis import given
+        import hypothesis.strategies as st
+
+        pytest_plugins = 'pytest_asyncio'
+
+        @given(n=st.integers())
+        async def test_hypothesis(n: int):
+            assert isinstance(n, int)
+        """
+        )
+    )
+    result = pytester.runpytest("--asyncio-mode=auto")
+    result.assert_outcomes(passed=1)
+
+
+def test_sync_not_auto_marked(pytester):
+    """Assert that synchronous Hypothesis functions are not marked with asyncio"""
+    pytester.makepyfile(
+        dedent(
+            """\
+        import asyncio
+        import pytest
+        from hypothesis import given
+        import hypothesis.strategies as st
+
+        pytest_plugins = 'pytest_asyncio'
+
+        @given(n=st.integers())
+        def test_hypothesis(request, n: int):
+            markers = [marker.name for marker in request.node.own_markers]
+            assert "asyncio" not in markers
+            assert isinstance(n, int)
+        """
+        )
+    )
+    result = pytester.runpytest("--asyncio-mode=auto")
+    result.assert_outcomes(passed=1)

